### PR TITLE
Record event and metric for lifecycle fallback to http

### LINKED ISSUE #113174: Add metrics and an event when lifecycle handlers fallback to http
### What would you like to be added?

Emit an event and increment a metric when the lifecycle handler for an https action falls back to http as @liggitt requested in #86139.

### Why is this needed?

To allow cluster users to identify lifecycle handlers that are configured to speak https to an http server.

### PR DESCRIPTION
**Pull request by liggitt**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Follow up to https://github.com/kubernetes/kubernetes/pull/86139 that adds event and metric recording when a https lifecycle hook falls back to http

#### Which issue(s) this PR fixes:

fixes https://github.com/kubernetes/kubernetes/issues/113174

#### Special notes for your reviewer:

Builds on https://github.com/kubernetes/kubernetes/pull/86139, only the last two commits are new

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/cc @tallclair @bhcleek

### CODE REVIEW

File: pkg/kubelet/lifecycle/handlers.go [hunk 1]
```diff
@@ -149,6 +152,10 @@ func (hr *handlerRunner) runHTTPHandler(pod *v1.Pod, container *v1.Container, ha
 
 			// clear err since the fallback succeeded
 			if httpErr == nil {
+				if eventRecorder != nil {
```
Comment (tallclair) [inline pkg/kubelet/lifecycle/handlers.go:155]: Delete todo at line 144
```
// TODO: emit an event about the fallback
```

Can you move the metric TODO down to here while you're at it?

File: pkg/kubelet/lifecycle/handlers.go [hunk 2]
```diff
@@ -149,6 +152,10 @@ func (hr *handlerRunner) runHTTPHandler(pod *v1.Pod, container *v1.Container, ha
...
+					// report the fallback with an event
+					eventRecorder.Event(pod, v1.EventTypeWarning, "LifecycleHTTPFallback", fmt.Sprintf("request to HTTPS lifecycle hook %s got HTTP response, retry with HTTP succeeded", req.URL.Host))
```
Comment (tallclair) [inline pkg/kubelet/lifecycle/handlers.go:157]: Do we know whether this is the PostStart or PreStop handler at this point? That might be useful to include in the event if so.

**Review by tallclair — COMMENTED** [review]
/hold

For https://github.com/kubernetes/kubernetes/pull/86139 to merge

File: pkg/kubelet/lifecycle/handlers.go (hunk 1 above)
Comment (liggitt) [reply pkg/kubelet/lifecycle/handlers.go:154]: done

File: pkg/kubelet/lifecycle/handlers.go (hunk 2 above)
Comment (liggitt) [reply pkg/kubelet/lifecycle/handlers.go:156]: we don't, but the port number should be sufficient to identify the problematic endpoint

File: pkg/kubelet/lifecycle/handlers.go [hunk 3]
```diff
@@ -149,6 +150,11 @@ func (hr *handlerRunner) runHTTPHandler(pod *v1.Pod, container *v1.Container, ha
 
 			// clear err since the fallback succeeded
 			if httpErr == nil {
+				// TODO: increment a metric about the fallback
+				if eventRecorder != nil {
+					// report the fallback with an event
+					eventRecorder.Event(pod, v1.EventTypeWarning, "LifecycleHTTPFallback", fmt.Sprintf("request to HTTPS lifecycle hook %s got HTTP response, retry with HTTP succeeded", req.URL.Host))
```
Comment (bhcleek) [inline pkg/kubelet/lifecycle/handlers.go:156]: Should the log at https://github.com/kubernetes/kubernetes/pull/113175/commits/122b43037eccc0c17ecd912e22d8e9f92a091040#diff-66f9167fad0ae80b15631b755bc6f6b3ce9324eec823ec01dd47d257707e9870R144 be changed to WARN level to maintain consistency with the event type?

Comment (liggitt) [reply pkg/kubelet/lifecycle/handlers.go:156]: there's no WarnS, only InfoS and ErrorS

**Comment by liggitt** [comment]
added a commit to record a counter metric as well.

/assign @logicalhan 
to look at the metric commit

File: pkg/kubelet/metrics/metrics.go [hunk 4]
```diff
@@ -86,6 +86,9 @@ const (
 	// Metrics to track ephemeral container usage by this kubelet
 	ManagedEphemeralContainersKey = "managed_ephemeral_containers"
 
+	// Metrics to track total number of lifecycle handler calls that fell back to http from https
+	LifecycleHandlerHTTPFallbacksKey = "lifecycle_handler_http_fallbacks_total"
```
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/kubelet/metrics/metrics.go]: thanks, updated

**Review by logicalhan — COMMENTED** [review]
/lgtm
/approve
(from sig inst)

**Comment by pacoxu** [comment]
/hold cancel
as #86139 was merged

**Pull request merged**